### ISSUE TITLE
Bug in decode_json_fields: data loss, type conflict

### ISSUE THREAD

**Issue by f-ld** [elastic/beats#3534]
Problem first exposed here: https://discuss.elastic.co/t/bug-in-decode-json-fields-data-loss-tye-conflict/73812/2

I think the following behaviour is a bug because it causes some data loss.

Please assume the following log file `/tmp/sample_json.log`:
```
{"someDate":"2016-09-28T01:40:26.760+0000", "someNumberAsString": "1475026826760", "someNumber": 1475026826760, "someString": "foobar", "someString2": "2017 is awesome"}
```

Now the following filebeat 5.2 configuration:
```
filebeat.prospectors:
- input_type: log
  paths:
    - /tmp/sample_json.log

processors:
- decode_json_fields:
    fields: ["message"]
    target: "foobar"
    max_depth: 10

output.logstash:
  hosts: ["localhost:5044"]
```

The this is the decoded output in logstash ruby debug output:
```
{
       "message" => "{\"someDate\":\"2016-09-28T01:40:26.760+0000\", \"someNumberAsString\": \"1475026826760\", \"someNumber\": 1475026826760, \"someString\": \"foobar\", \"someString2\": \"2017 is awesome\"}",
      "@version" => "1",
    "@timestamp" => "2017-02-03T08:53:57.579Z",
          "beat" => {
            "name" => "fabien1",
        "hostname" => "fabien1",
         "version" => "5.2.0"
    },
        "source" => "/tmp/sample_json.log",
        "offset" => 170,
        "foobar" => {
                  "someDate" => 2016,
                "someNumber" => 1475026826760,
        "someNumberAsString" => 1475026826760,
                "someString" => "foobar",
               "someString2" => 2017
    },
          "type" => "log",
    "input_type" => "log",
          "host" => "fabien1",
          "tags" => [
        [0] "beats_input_codec_plain_applied"
    ]
}
```

The problem here is that we loose the fact that `someString2` was not only a number but a full string:
- so if for other records it was "What an awesome year!" then we will have a type conflict: sometimes an int, sometimes a string
- same for "someDate" that becomes `2016` with a lot of data loss
In conclusion, this parsing breaks the schema of the data: only json fields that were numbers should be converted as numbers. 

I agree that changing this behaviour might break some compatibility for people assuming to get integers whenever possible, but the default behaviour is unusable in most cases.

**Comment by andrewkroh** [comment]
This test case passes so maybe the problem is outside of `decode_json_fields`. Not sure yet.

```
func TestDecodeJSONNumbers(t *testing.T) {
	msg := `{
  "someDate": "2016-09-28T01:40:26.760+0000",
  "someNumberAsString": "1475026826760",
  "someNumber": 1475026826760,
  "someString": "foobar",
  "someString2": "2017 is awesome"
}`

	input := common.MapStr{
		"msg": msg,
	}

	config, err := common.NewConfigFrom(map[string]interface{}{
		"fields":        fields,
		"target":        "foobar",
		"max_depth":     10,
	})
	if err != nil {
		t.Fatal(err)
	}

	actual := getActualValue(t, config, input)
	decoded, ok := actual["foobar"].(map[string]interface{})
	if !ok {
		t.Fatal("msg in not a dict")
	}
	t.Log(actual)

	assert.Equal(t, "2016-09-28T01:40:26.760+0000", decoded["someDate"])
	assert.EqualValues(t, "1475026826760", decoded["someNumberAsString"])
	assert.EqualValues(t, 1475026826760, decoded["someNumber"])
	assert.Equal(t, "foobar", decoded["someString"])
	assert.Equal(t, "2017 is awesome", decoded["someString2"])
}
```

**Comment by andrewkroh** [comment]
I can reproduce the issue with the file output so it's not specific to Logstash.

```
{
  "@timestamp": "2017-02-07T07:57:25.757Z",
  "beat": {
    "hostname": "host",
    "name": "host",
    "version": "5.2.0"
  },
  "foobar": {
    "someDate": 2016,
    "someNumber": 1475026826760,
    "someNumberAsString": 1475026826760,
    "someString": "foobar",
    "someString2": 2017
  },
  "input_type": "log",
  "message": "{\"someDate\":\"2016-09-28T01:40:26.760+0000\", \"someNumberAsString\": \"1475026826760\", \"someNumber\": 1475026826760, \"someString\": \"foobar\", \"someString2\": \"2017 is awesome\"}",
  "offset": 170,
  "source": "input.json",
  "type": "log"
}
```

**Comment by andrewkroh** [comment]
@f-ld Can you try setting `max_depth: 1` in your config file?

**Comment by f-ld** [comment]
@andrewkroh : I can confirm the behaviour is different:

- `max_depth: 2` (and more):
```
        "foobar" => {
                  "someDate" => 2016,
                "someNumber" => 1475026826760,
        "someNumberAsString" => 1475026826760,
                "someString" => "foobar",
               "someString2" => 2017
    },
```
- `max_depth:1`:
```
        "foobar" => {
                  "someDate" => "2016-09-28T01:40:26.760+0000",
                "someNumber" => 1475026826760,
        "someNumberAsString" => "1475026826760",
                "someString" => "foobar",
               "someString2" => "2017 is awesome"
    },
```

**Comment by urso** [comment]
Looks like user error to me. 1) why use the `decode_json_fields` processor? The prospector can already parse json 2) the `decode_json_fields` processor can recursively parse string content... like whenever we find a string, check if it's a valid json and try to parse. This is, cause some loggers (e.g. docker...) or tools collecting logs might include json content as string... By setting `max_depth>1` you did enforce recursive parsing, which would find `someNumberAsString` being a json number.

**Comment by f-ld** [comment]
I tried to use that feature because it gives some flexibility that the json parsing of the prospector does not offer (See https://github.com/elastic/beats/issues/3025).
And while testing it I was surprised to see that string being a number or starting by a number were converted to a number (dropping any other content when it exists). Reason why I asked about it in the forum and opened this issue when suggested there.

**Comment by andrewkroh** [comment]
I expected that the recursive decoding would only operate on objects and arrays. That's why I was surprised to see it modifying strings that are just numbers.

I think the recursive decoding would be more useful if it only operated on objects and arrays. Maybe we could add a cursory check to see if the string begins with a `{` or `[` before attempting to decode it.

**Comment by anonimnor** [comment]
this elastic piece of sh#t converts not only strings starting with numbers into numbers, but also strings starting with booleans to booleans, - wich is damn wise, but, at the same time, then, it does not represent 1 or '1' as True boolean for example. the boolean overplacement will fail if we feed em '1 shit' or 'true shit', or true, or whatever, then try to push a string or number there.
gj elastic arses go fu with your junk.
